src/service/batched-xhr-impl.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Xhr} from './xhr-impl';
+import {getService, registerServiceBuilder} from '../service';
+import {map} from '../utils/object';
+import {removeFragment} from '../url';
+
+
+/**
+ * A wrapper around the Xhr service which batches the result of GET requests
+ *
+ * @package Visible for type.
+ * @visibleForTesting
+ */
+export class BatchedXhr extends Xhr {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    super(win);
+
+    /** @const {!Object<!Promise<!./xhr-impl.FetchResponse>>} */
+    this.fetchPromises_ = map();
+  }
+
+  /**
+   * Fetch and batch the requests if possible.
+   *
+   * @param {string} input URL
+   * @param {?./xhr-impl.FetchInitDef=} opt_init Fetch options object.
+   * @return {!Promise<!./xhr-impl.FetchResponse>}
+   * @override
+   */
+  fetch(input, opt_init) {
+    const accept =
+        (opt_init && opt_init.headers && opt_init.headers['Accept']) || '';
+    const isBatchable =
+        (!opt_init || !opt_init.method || opt_init.method === 'GET');
+    const key = this.getMapKey_(input, accept);
+    const isBatched = !!this.fetchPromises_[key];
+
+    if (isBatchable && isBatched) {
+      return this.fetchPromises_[key].then(response => response.clone());
+    }
+
+    const fetchPromise = super.fetch(input, opt_init);
+
+    if (isBatchable) {
+      this.fetchPromises_[key] = fetchPromise.then(response => {
+        delete this.fetchPromises_[key];
+        return response.clone();
```

Comment:
This isn't sufficient to prevent double draining. The problem is this is part of the promise chain.
So, first call to fetch returns the clone. I then add to the chain to read the body, and issue a second request to the same endpoint. The second request will see the batched call, and return a clone of that promise (a clone of a clone) Problem is, that clone will happen _after_ the body read. (Because execution chains on a single promise is defined)

This should return the response without cloning. This total promise chain will be stored as the batch. For the return value, we should return the batching promise with a `.then((resp) => resp.clone())` added to its chain.